kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/KuraException.java

```diff
@@ -20,167 +18,191 @@
-				messagePattern = resourceBundle.getString(code.name());
-				if (messagePattern == null) {
-					s_logger.warn("Could not find Exception Messages for Locale {} and code {}", locale, code);
-				}
-			}
-		} catch (MissingResourceException mre) {
-			// log the failure to load a message bundle
-			s_logger.warn("Could not load Exception Messages Bundle for Locale {}", locale);
-		}
-
-		//
-		// If no bundle or code in the bundle is found, use a generic message
-		if (messagePattern == null) {
-			if (code != null) {
-				// build a generic message format
-				messagePattern = MessageFormat.format(KURA_GENERIC_MESSAGES_PATTERN, code.name());
-			}
-			else {
-				// build a generic message format
-				messagePattern = MessageFormat.format(KURA_GENERIC_MESSAGES_PATTERN, "Unknown");			
-			}
-		}
-
-		return messagePattern;
-	}
+public class KuraException extends Exception {
+
+    /** The Constant denoting resource bundle. */
+    private static final String KURA_EXCEPTION_MESSAGES_BUNDLE = "org.eclipse.kura.core.messages.KuraExceptionMessagesBundle";
+
+    /** The Constant denoting message pattern. */
+    private static final String KURA_GENERIC_MESSAGES_PATTERN = "Generic Error - {0}: {1}";
+
+    /** The Logger. */
+    private static final Logger s_logger = LoggerFactory.getLogger(KuraException.class);
+
+    /** The Constant denoting serial version identifier. */
+    private static final long serialVersionUID = 7468633737373095296L;
+
+    /** The associated arguments. */
+    private transient Object[] arguments;
+
+    /** The associated error code. */
+    private final KuraErrorCode code;
+
+    /**
+     * Builds a new {@link KuraException} instance based on the supplied {@link KuraErrorCode}.
+     *
+     * @param code
+     *            the error code
+     */
+    public KuraException(final KuraErrorCode code) {
+        this.code = code;
+    }
+
+    /**
+     * Builds a new {@link KuraException} instance based on the supplied {@link KuraErrorCode}.
+     *
+     * @param code
+     *            the error code
+     * @param arguments
+     *            the arguments
+     */
+    public KuraException(final KuraErrorCode code, final Object... arguments) {
+        this.code = code;
+        this.arguments = arguments;
+    }
+
+    /**
+     * Builds a new {@link KuraException} instance based on the supplied {@link KuraErrorCode}, an optional Throwable
+     * cause, and optional arguments for the associated exception message.
+     *
+     * @param code
+     *            the error code
+     * @param cause
+     *            the cause
+     * @param arguments
+     *            the arguments
+     */
+    public KuraException(final KuraErrorCode code, final Throwable cause, final Object... arguments) {
+        super(cause);
+        this.code = code;
+        this.arguments = arguments;
+    }
+
+    /**
+     * Factory method to build an {@link KuraException} with the {@link KuraErrorCode#INTERNAL_ERROR} code providing
```

Comment:
There should be at least a comment to not use this one. As we agreed on not using "INTERNAL_ERROR" anymore.